#  Exploratory Data Analysis for the Kaggle Airline Delays dataset
## Introduction
This dataset contains information on **airline delays in the United States** for **December 2019 and December 2020**.  
The focus is on arrival flights, cancellations, diversions, and delays, along with their causes (carrier, weather, airspace system, security, and late aircraft).  

##  Description
- **Timeframe**: December 2019 and December 2020  
- **Scope**: U.S. domestic flights  
- **Unit of analysis**: Flights aggregated **per carrier per U.S. city**  
- **Features included**:  
  - `arr_flights` — total arrival flights  
  - `carrier_ct` — number of carrier-related delays  
  - `weather_ct` — number of weather-related delays  
  - `nas_ct` — number of NAS/system delays  
  - `security_ct` — number of security-related delays  
  - `late_aircraft_ct` — number of late aircraft delays  
  - `arr_cancelled` — cancelled flights  
  - `arr_diverted` — diverted flights  


The dataset allows us to explore delay patterns, identify the main contributors to disruptions, and compare performance across carriers and cities.  
Source: Kaggle, Airline Delays CSV file.


In [85]:
# -------------------------------
# Import necessary libraries
# -------------------------------
import pandas as pd
import numpy as np
import os

# # -----Load the dataset -----
df_airline_delays= pd.read_csv("airline_delay.csv")

# ----- Data set display -----
print("═" * 50)
print("FIRST 10 ROWS OF DATASET:")
print("═" * 50)
df_airline_delays.head(10)


══════════════════════════════════════════════════
FIRST 10 ROWS OF DATASET:
══════════════════════════════════════════════════


year  month carrier       carrier_name airport  \
0  2020     12      9E  Endeavor Air Inc.     ABE   
1  2020     12      9E  Endeavor Air Inc.     ABY   
2  2020     12      9E  Endeavor Air Inc.     AEX   
3  2020     12      9E  Endeavor Air Inc.     AGS   
4  2020     12      9E  Endeavor Air Inc.     ALB   
5  2020     12      9E  Endeavor Air Inc.     ATL   
6  2020     12      9E  Endeavor Air Inc.     ATW   
7  2020     12      9E  Endeavor Air Inc.     AVL   
8  2020     12      9E  Endeavor Air Inc.     AZO   
9  2020     12      9E  Endeavor Air Inc.     BDL   

                                        airport_name  arr_flights  arr_del15  \
0  Allentown/Bethlehem/Easton, PA: Lehigh Valley ...         44.0        3.0   
1             Albany, GA: Southwest Georgia Regional         90.0        1.0   
2           Alexandria, LA: Alexandria International         88.0        8.0   
3        Augusta, GA: Augusta Regional at Bush Field        184.0        9.0   
4                   Albany, NY: Albany International         76.0       11.0   
5  Atlanta, GA: Hartsfield-Jackson Atlanta Intern...       5985.0      445.0   
6               Appleton, WI: Appleton International        142.0       14.0   
7                  Asheville, NC: Asheville Regional        147.0       10.0   
8  Kalamazoo, MI: Kalamazoo/Battle Creek Internat...         84.0       14.0   
9                Hartford, CT: Bradley International        150.0       19.0   

   carrier_ct  weather_ct  ...  security_ct  late_aircraft_ct  arr_cancelled  \
0        1.63        0.00  ...          0.0              1.25            0.0   
1        0.96        0.00  ...          0.0              0.00            0.0   
2        5.75        0.00  ...          0.0              0.65            0.0   
3        4.17        0.00  ...          0.0              3.00            0.0   
4        4.78        0.00  ...          0.0              1.00            1.0   
5      142.89       11.96  ...          1.0            127.79            5.0   
6        5.36        0.00  ...          0.0              0.94            1.0   
7        6.04        1.00  ...          0.0              1.96            0.0   
8        6.24        0.96  ...          0.0              0.00            1.0   
9        5.70        0.00  ...          0.0              1.23            3.0   

   arr_diverted  arr_delay  carrier_delay  weather_delay  nas_delay  \
0           1.0       89.0           56.0            0.0        3.0   
1           0.0       23.0           22.0            0.0        1.0   
2           1.0      338.0          265.0            0.0       45.0   
3           0.0      508.0          192.0            0.0       92.0   
4           0.0      692.0          398.0            0.0      178.0   
5           0.0    30756.0        16390.0         1509.0     5060.0   
6           0.0      436.0          162.0            0.0      182.0   
7           1.0     1070.0          838.0          141.0       24.0   
8           1.0     2006.0         1164.0          619.0      223.0   
9           0.0      846.0          423.0            0.0      389.0   

   security_delay  late_aircraft_delay  
0             0.0                 30.0  
1             0.0                  0.0  
2             0.0                 28.0  
3             0.0                224.0  
4             0.0                116.0  
5            16.0               7781.0  
6             0.0                 92.0  
7             0.0                 67.0  
8             0.0                  0.0  
9             0.0                 34.0  

[10 rows x 21 columns]

In [86]:
# -------------------------------
# Basic information about the dataset
# -------------------------------
print("\n" + "═" * 50)
print("BASIC INFORMATION ABOUT THE DATASET:")
print("═" * 50)

print(f"\nDataset dimensions: {df_airline_delays.shape}")
print(f"\nNumber of rows: {df_airline_delays.shape[0]}")
print(f"\nNumber of columns: {df_airline_delays.shape[1]}")




══════════════════════════════════════════════════
BASIC INFORMATION ABOUT THE DATASET:
══════════════════════════════════════════════════

Dataset dimensions: (3351, 21)

Number of rows: 3351

Number of columns: 21


In [87]:
# -------------------------------
# Information on data types
# -------------------------------

print("\n" + "═" * 50)
print("INFORMATION ON DATA TYPES:")
print("═" * 50)
df_airline_delays.info()



══════════════════════════════════════════════════
INFORMATION ON DATA TYPES:
══════════════════════════════════════════════════
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3351 entries, 0 to 3350
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   year                 3351 non-null   int64  
 1   month                3351 non-null   int64  
 2   carrier              3351 non-null   object 
 3   carrier_name         3351 non-null   object 
 4   airport              3351 non-null   object 
 5   airport_name         3351 non-null   object 
 6   arr_flights          3343 non-null   float64
 7   arr_del15            3343 non-null   float64
 8   carrier_ct           3343 non-null   float64
 9   weather_ct           3343 non-null   float64
 10  nas_ct               3343 non-null   float64
 11  security_ct          3343 non-null   float64
 12  late_aircraft_ct     3343 non-null   float64
 13  arr_canc

In [88]:
print(df_airline_delays.columns.tolist())


['year', 'month', 'carrier', 'carrier_name', 'airport', 'airport_name', 'arr_flights', 'arr_del15', 'carrier_ct', 'weather_ct', 'nas_ct', 'security_ct', 'late_aircraft_ct', 'arr_cancelled', 'arr_diverted', 'arr_delay', 'carrier_delay', 'weather_delay', 'nas_delay', 'security_delay', 'late_aircraft_delay']


In [89]:
# -------------------------------
# Descriptive statistics
# -------------------------------
print("\n" + "═" * 50)
print("DESCRIPTIVE STATISTICS:")
print("═" * 50)
print(df_airline_delays.describe().round(2))


══════════════════════════════════════════════════
DESCRIPTIVE STATISTICS:
══════════════════════════════════════════════════
          year   month  arr_flights  arr_del15  carrier_ct  weather_ct  \
count  3351.00  3351.0      3343.00    3343.00     3343.00     3343.00   
mean   2019.46    12.0       298.27      51.00       16.07        1.44   
std       0.50     0.0       852.44     146.48       41.76        4.82   
min    2019.00    12.0         1.00       0.00        0.00        0.00   
25%    2019.00    12.0        35.00       5.00        1.49        0.00   
50%    2019.00    12.0        83.00      12.00        4.75        0.06   
75%    2020.00    12.0       194.50      33.00       12.26        1.01   
max    2020.00    12.0     19713.00    2289.00      697.00       89.42   

        nas_ct  security_ct  late_aircraft_ct  arr_cancelled  arr_diverted  \
count  3343.00      3343.00           3343.00        3343.00       3343.00   
mean     16.18         0.14             17.17     

# Descriptive Statistics Analysis of Flight Delays

## General Overview
- **Year**: Data only covers **2019 and 2020**.  
- **Month**: All records are for **December (12)** → dataset is focused only on December 2019 and December 2020.  
- **Count**: ~3,351 rows → solid sample size.  

---

## Flights
- **arr_flights (arrival flights)**  
  - **Mean**: ~298  
  - **Median (50%)**: 83  
  - **Max**: 19,713  
  - ➝ A few very large airports dominate, while most airports handle <100 flights.  

- **arr_cancelled (canceled flights)**  
  - **Median**: 0  
  - **Max**: 224  
  - ➝ Most airports had no cancellations, but some outliers show large numbers.  

- **arr_diverted (diverted flights)**  
  - **Median**: 0  
  - **Max**: 42  
  - ➝ Diversions are rare.  

---

## Delays and Causes
- **arr_del15 (flights delayed >15 min)**  
  - **Mean**: 51  
  - **Median**: 12  
  - **Max**: 2,289  
  - ➝ Most airports have relatively few delays, but some airports have thousands.  

- **Delay causes (counts of delayed flights):**
  - **Carrier_ct (airline responsibility)** → mean 16, **max ~697**  
  - **Late_aircraft_ct (previous aircraft delayed)** → mean 17, **max ~820**  
  - **NAS_ct (airspace system issues)** → mean 16, **max ~1,039**  
  - **Weather_ct** → mean 1.4, **max ~89**  
  - **Security_ct** → mean 0.14, **max ~17**  

**Conclusion**: The top three causes are **airline responsibility, late aircraft, and NAS issues**. Weather and security play a much smaller role.  

---

##  Length of Delays
- **arr_delay (total arrival delay minutes)**  
  - **Mean**: 1.25 minutes  
  - **Median**: 0 minutes  
  - **Max**: 1,150 minutes  
  - ➝ Most flights are on time; averages are inflated by outliers.  

- **Other delays (minutes):**  
  - **Carrier delay, weather delay, NAS delay, late aircraft delay** → mostly 0, but extreme outliers exist.  
  - **Security delay** → nearly always 0.  

---

## Key Insights
1. **Highly skewed distribution** → Most flights/airports experience little to no delay, but a few extreme cases distort averages.  
2. **Main drivers of delays**:  
   - Airline responsibility  
   - Late aircraft from previous flights  
   - NAS (air traffic system) issues  
3. **Weather and security delays are rare**, though extreme weather can cause spikes.  
4. **Cancellations and diversions are uncommon**, but some airports show clusters of cancellations.  
5. **Median values (50%) give a clearer picture** than means, since the dataset is dominated by outliers.  


In [90]:
# -----------------------------------------------
# Check unique values for categorical columns
# -----------------------------------------------
print("UNIQUE VALUES FOR CATEGORICAL COLUMNS:")
print("═" * 50)
print("Number of Unique Carriers:", df_airline_delays['carrier'].nunique())
print("Number of Unique Airports:", df_airline_delays['airport'].nunique())
print("Months in Data:", df_airline_delays['month'].unique())
print("Years in Data:", df_airline_delays['year'].unique())


UNIQUE VALUES FOR CATEGORICAL COLUMNS:
══════════════════════════════════════════════════
Number of Unique Carriers: 17
Number of Unique Airports: 360
Months in Data: [12]
Years in Data: [2020 2019]


In [91]:
# -------------------------------
# Missing values in dataset
# -------------------------------
print("\n" + "═" * 50)
print("MISSING VALUES IN DATASET:")
print("═" * 50)
missing_values = df_airline_delays.isnull().sum()
missing_percent = (df_airline_delays.isnull().sum() / len(df_airline_delays)) * 100
missing_df = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage (%)': missing_percent
})
print(missing_df[missing_df['Missing Values'] > 0])


══════════════════════════════════════════════════
MISSING VALUES IN DATASET:
══════════════════════════════════════════════════
                     Missing Values  Percentage (%)
arr_flights                       8        0.238735
arr_del15                         8        0.238735
carrier_ct                        8        0.238735
weather_ct                        8        0.238735
nas_ct                            8        0.238735
security_ct                       8        0.238735
late_aircraft_ct                  8        0.238735
arr_cancelled                     8        0.238735
arr_diverted                      8        0.238735
arr_delay                         8        0.238735
carrier_delay                     8        0.238735
weather_delay                     8        0.238735
nas_delay                         8        0.238735
security_delay                    8        0.238735
late_aircraft_delay               8        0.238735


In [92]:
# --------------------------------------------------------------
# Cleaning of numerical columns and filling with median values
# ---------------------------------------------------------------
df_clean = df_airline_delays.copy()
numeric_cols = df_airline_delays.select_dtypes(include=[np.number]).columns.tolist()
for col in numeric_cols:
    if df_airline_delays[col].isnull().any():
        df_clean[col] = df_airline_delays[col].fillna(df_airline_delays[col].median())

# ----- Cleaning of categorical columns -----         

df_clean = df_airline_delays.dropna(subset=[
    "arr_flights", "arr_del15", "carrier_ct", "weather_ct", 
    "nas_ct", "security_ct", "late_aircraft_ct", "arr_cancelled", 
    "arr_diverted", "arr_delay", "carrier_delay", "weather_delay", 
    "nas_delay", "security_delay", "late_aircraft_delay"
])

In [93]:
# -----------------------------------------------
# Checking for NaN values ​​after cleanup
# -----------------------------------------------
print("\n" + "═" * 50)
print("CHECKING FOR NAN VALUES AFTER CLEANUP:")
print("═" * 50)

missing_after = df_clean.isnull().sum()
print(missing_after)


══════════════════════════════════════════════════
CHECKING FOR NAN VALUES AFTER CLEANUP:
══════════════════════════════════════════════════
year                   0
month                  0
carrier                0
carrier_name           0
airport                0
airport_name           0
arr_flights            0
arr_del15              0
carrier_ct             0
weather_ct             0
nas_ct                 0
security_ct            0
late_aircraft_ct       0
arr_cancelled          0
arr_diverted           0
arr_delay              0
carrier_delay          0
weather_delay          0
nas_delay              0
security_delay         0
late_aircraft_delay    0
dtype: int64


In [94]:
print(df_clean.columns)


Index(['year', 'month', 'carrier', 'carrier_name', 'airport', 'airport_name',
       'arr_flights', 'arr_del15', 'carrier_ct', 'weather_ct', 'nas_ct',
       'security_ct', 'late_aircraft_ct', 'arr_cancelled', 'arr_diverted',
       'arr_delay', 'carrier_delay', 'weather_delay', 'nas_delay',
       'security_delay', 'late_aircraft_delay'],
      dtype='object')


In [95]:
# -----------------------------------------------
# Correlation matrix and heatmap visualization
# WITH MULTIPLE SAVE OPTIONS (no Chrome needed)
# -----------------------------------------------

import plotly.graph_objects as go
import os

# Create directory for visualizations if it doesn't exist
os.makedirs("visualizations", exist_ok=True)

# ----- Select numeric columns -----
numeric_cols = df_clean.select_dtypes(include='number')

# ----- Compute correlation -----
correlation_matrix = numeric_cols.corr()

# -----Filter strong correlations (0.7 < abs(corr) < 1) -----
abs_corr = correlation_matrix.abs()
top_cols = abs_corr[(abs_corr > 0.7) & (abs_corr < 1.0)] \
    .dropna(axis=0, how='all') \
    .dropna(axis=1, how='all') \
    .columns
filtered_corr = numeric_cols[top_cols].corr() if len(top_cols) > 0 else correlation_matrix

# ----- Create custom hover text -----
# Build a matrix of hover text strings
hover_text = []
for i, row_name in enumerate(filtered_corr.index):
    hover_row = []
    for j, col_name in enumerate(filtered_corr.columns):
        corr_value = filtered_corr.iloc[i, j]
        
        # Determine correlation strength
        abs_corr_val = abs(corr_value)
        if abs_corr_val == 1.0:
            strength = "Perfect"
        elif abs_corr_val >= 0.7:
            strength = "Strong"
        elif abs_corr_val >= 0.5:
            strength = "Moderate"
        elif abs_corr_val >= 0.3:
            strength = "Weak"
        else:
            strength = "Very Weak"
        
        # Determine direction
        if corr_value == 1.0:
            direction = " (identical)"
        elif corr_value > 0:
            direction = "Positive"
        elif corr_value < 0:
            direction = "Negative"
        else:
            direction = "No"
        
        # Create hover text
        if row_name == col_name:
            text = f"<b>{row_name}</b><br>" + \
                   f"Correlation with itself: 1.00<br>" + \
                   f"<i>Always perfectly correlated</i>"
        else:
            text = f"<b>{row_name}</b> vs <b>{col_name}</b><br>" + \
                   f"Correlation: {corr_value:.3f}<br>" + \
                   f"<b>{strength} {direction} Relationship</b><br>" + \
                   f"<i>{'Variables move together' if corr_value > 0 else 'Variables move oppositely' if corr_value < 0 else 'No linear relationship'}</i>"
        
        hover_row.append(text)
    hover_text.append(hover_row)

# ----- Plotly heatmap with custom hover -----
fig = go.Figure(data=go.Heatmap(
    z=filtered_corr.values,
    x=filtered_corr.columns,
    y=filtered_corr.index,
    text=filtered_corr.values,
    texttemplate='%{text:.2f}',
    textfont={"size": 10},
    colorscale='RdBu',
    zmin=-1,
    zmax=1,
    hovertemplate='%{customdata}<extra></extra>',
    customdata=hover_text,
    colorbar=dict(
        title="Correlation<br>Coefficient",
        tickmode="linear",
        tick0=-1,
        dtick=0.5,
        thickness=15,
        len=0.7,
        tickvals=[-1, -0.5, 0, 0.5, 1],
        ticktext=["-1.0<br>(Perfect<br>Negative)", 
                  "-0.5<br>(Moderate<br>Negative)", 
                  "0<br>(No<br>Correlation)", 
                  "0.5<br>(Moderate<br>Positive)", 
                  "1.0<br>(Perfect<br>Positive)"]
    )
))

# ----- Update layout for clarity -----
fig.update_layout(
    title={
        'text': "Correlation Matrix Heatmap<br><sub>Showing relationships between numeric variables</sub>",
        'x': 0.5,
        'xanchor': 'center',
        'y': 0.95,
        'yanchor': 'top'
    },
    xaxis=dict(
        title="Variables",
        title_standoff=30,  # Closer spacing
        tickangle=-40,      # Rotate labels to prevent overlap
        side="bottom"
    ),
    yaxis=dict(
        title="Variables",
        title_standoff=35  # Closer spacing
    ),
    width=1400,
    height=600,
    plot_bgcolor="white",
    paper_bgcolor="lightgray",
    hoverlabel=dict(
        bgcolor="white",
        font_size=13,
        font_family="Arial",
        bordercolor="black"
    ),
    font=dict(size=11),
    margin=dict(
        t=120,  # Top margin for title
        b=150,  # Bottom margin 
        l=80,   # Left margin
        r=80    # Right margin
    )
)

# ----- Add annotation explaining correlation -----
fig.add_annotation(
    text="Values range from -1 (perfect negative) to +1 (perfect positive correlation)",
    xref="paper", yref="paper",
    x=0.5, y=-0.3,  # Properly positioned
    showarrow=False,
    font=dict(size=10, color="gray"),
    xanchor='center'
)

# ----- Display the figure -----
fig.show()


fig.write_html("visualizations/correlation_heatmap.html")
print(" Saved interactive HTML: visualizations/correlation_heatmap.html")



 Saved interactive HTML: visualizations/correlation_heatmap.html


In [96]:
# -----------------------------------------
# Display of the strongest correlations
# ----------------------------------------
print("\n" + "═" * 50)
print("STRONGEST CORRELATIONS:")
print("═" * 50)


corr_pairs = correlation_matrix.unstack().sort_values(key=abs, ascending=False)
corr_pairs = corr_pairs[corr_pairs != 1.0]  
print(corr_pairs.head(10))


══════════════════════════════════════════════════
STRONGEST CORRELATIONS:
══════════════════════════════════════════════════
late_aircraft_ct     late_aircraft_delay    0.962579
late_aircraft_delay  late_aircraft_ct       0.962579
arr_delay            arr_del15              0.960250
arr_del15            arr_delay              0.960250
late_aircraft_ct     arr_del15              0.949273
arr_del15            late_aircraft_ct       0.949273
arr_delay            late_aircraft_delay    0.948945
late_aircraft_delay  arr_delay              0.948945
arr_flights          arr_del15              0.928553
arr_del15            arr_flights            0.928553
dtype: float64


In [97]:
# ----------------------------------------
# Define delay types for count analysis
# ----------------------------------------
delay_count_columns = ['carrier_ct', 'weather_ct', 'nas_ct', 'security_ct', 'late_aircraft_ct']

In [98]:
# ------------------------
# Summary statistics
# -------------------------

# ----- Ensure df_clean is a proper copy -----
df_clean = df_clean.copy()

# ----- Calculate delay rate ----- 
df_clean['delay_rate'] = df_clean['arr_del15'] / df_clean['arr_flights']


# ----- Calculate delay rate----- 
df_clean['delay_rate'] = df_clean['arr_del15'] / df_clean['arr_flights']

# ----- Airlines with highest delay rate (with minimum 100 flights for significance)----- 
significant_airlines = df_clean[df_clean['arr_flights'] >= 100]
airline_delay_rates = significant_airlines.groupby('carrier_name')['delay_rate'].mean().sort_values(ascending=False)

print("Airlines with Highest Delay Rates (min 100 flights):")
print(airline_delay_rates.head(10))

# ----- Airports with highest delay rate ----- 
significant_airports = df_clean[df_clean['arr_flights'] >= 100]
airport_delay_rates = significant_airports.groupby('airport_name')['delay_rate'].mean().sort_values(ascending=False)

print("\nAirports with Highest Delay Rates (min 100 flights):")
print(airport_delay_rates.head(10))


# ----- Define the correct delay count columns ----- 
delay_count_columns = ['carrier_ct', 'weather_ct', 'nas_ct', 'security_ct', 'late_aircraft_ct']

# ----- Most common delay type by airline ----- 
delay_by_airline = df_clean.groupby('carrier_name')[delay_count_columns].sum()
delay_by_airline['total_delays'] = delay_by_airline.sum(axis=1)

for delay_type in delay_count_columns:
    delay_by_airline[f'{delay_type}_pct'] = delay_by_airline[delay_type] / delay_by_airline['total_delays']

print("\nPrimary Delay Type by Airline:")
for airline in delay_by_airline.index:
    pct_columns = [f'{dt}_pct' for dt in delay_count_columns]
    primary_delay = delay_by_airline.loc[airline, pct_columns].idxmax()
    primary_delay = primary_delay.replace('_pct', '')
    print(f"{airline}: {primary_delay}")

# ----- Summary statistics ----- 
print(f"\nTotal Flights: {df_clean['arr_flights'].sum():,}")
print(f"Total Delayed Flights: {df_clean['arr_del15'].sum():,}")
print(f"Overall Delay Rate: {df_clean['arr_del15'].sum()/df_clean['arr_flights'].sum()*100:.2f}%")


Airlines with Highest Delay Rates (min 100 flights):
carrier_name
JetBlue Airways            0.273524
Allegiant Air              0.264285
ExpressJet Airlines LLC    0.248437
Mesa Airlines Inc.         0.237746
PSA Airlines Inc.          0.214595
Frontier Airlines Inc.     0.188769
Alaska Airlines Inc.       0.178865
Envoy Air                  0.171796
Southwest Airlines Co.     0.167964
United Air Lines Inc.      0.165303
Name: delay_rate, dtype: float64

Airports with Highest Delay Rates (min 100 flights):
airport_name
Concord, NC: Concord Padgett Regional                                     0.418605
Redding, CA: Redding Municipal                                            0.322581
Bakersfield, CA: Meadows Field                                            0.314571
Aguadilla, PR: Rafael Hernandez                                           0.298507
Akron, OH: Akron-Canton Regional                                          0.295359
Trenton, NJ: Trenton Mercer                                

In [99]:
# ------------------------------------------------------
# Top airlines and airports interactive visualization
# ------------------------------------------------------

# ----- Import necessary libraries -----
import plotly.graph_objects as go
import os

# ----- Create directory for visualizations -----
os.makedirs("visualizations", exist_ok=True)

# ----- Calculate delay rates first -----
# Group by airline and calculate delay rate
airline_stats = df_clean.groupby('carrier_name').agg({
    'arr_flights': 'sum',
    'arr_del15': 'sum'
}).reset_index()
airline_delay_rates = (airline_stats['arr_del15'] / airline_stats['arr_flights'])
airline_delay_rates.index = airline_stats['carrier_name']

# Group by airport and calculate delay rate
airport_stats = df_clean.groupby('airport_name').agg({
    'arr_flights': 'sum',
    'arr_del15': 'sum'
}).reset_index()
airport_delay_rates = (airport_stats['arr_del15'] / airport_stats['arr_flights'])
airport_delay_rates.index = airport_stats['airport_name']

# ----- Prepare data -----
top_airlines = airline_delay_rates.sort_values(ascending=False)
top_airports = airport_delay_rates.sort_values(ascending=False)

# ----- Plot top airlines -----
fig_airlines = go.Figure()

# Create custom hover text for airlines
hover_text_airlines = [f"<b>{airline}</b><br>" +
                       f"Delay Rate: {rate:.1%}<br>" +
                       f"<i>({rate*100:.1f} out of 100 flights delayed)</i>"
                       for airline, rate in zip(top_airlines.index[:10], 
                                               top_airlines.values[:10])]

fig_airlines.add_trace(go.Bar(
    x=top_airlines.values[:10],
    y=top_airlines.index[:10],
    orientation='h',
    marker=dict(color=top_airlines.values[:10], colorscale='Reds'),
    name='Airlines',
    hovertemplate='%{hovertext}<extra></extra>',  # Custom hover template
    hovertext=hover_text_airlines
))

# ----- Dropdown buttons for top N airlines -----
buttons_airlines = []
max_airlines = min(20, len(top_airlines))

for n in range(1, max_airlines + 1):
    # Create hover text for each dropdown option
    hover_text_n = [f"<b>{airline}</b><br>" +
                   f"Delay Rate: {rate:.1%}<br>" +
                   f"<i>({rate*100:.1f} out of 100 flights delayed)</i>"
                   for airline, rate in zip(top_airlines.index[:n], 
                                           top_airlines.values[:n])]
    
    buttons_airlines.append(dict(
        label=f"Top {n}",
        method="update",
        args=[{"x": [top_airlines.values[:n]],
               "y": [top_airlines.index[:n]],
               "marker.color": [top_airlines.values[:n]],
               "hovertext": [hover_text_n]},
              {"title": f"Top {n} Airlines by Delay Rate",
               "xaxis": {"title": "Delay Rate (%)", "tickformat": ".1%"},
               "yaxis": {"title": "Airline", "autorange": "reversed"}}]
    ))

fig_airlines.update_layout(
    title="Top 10 Airlines by Delay Rate",
    xaxis=dict(title="Delay Rate (%)", tickformat=".1%"),  # Format x-axis as percentage
    yaxis=dict(title="Airline", autorange="reversed"),
    plot_bgcolor='lightgray',
    paper_bgcolor='lightgray',
    width=1400,        
    height=400,
    hoverlabel=dict(
        bgcolor="white",
        font_size=14,
        font_family="Arial"
    ),
    updatemenus=[dict(
        active=9,  # default top 10
        buttons=buttons_airlines,
        direction="down",
        showactive=True,
        x=0,
        y=1.15
    )]
)

# ----- Plot top airports -----
fig_airports = go.Figure()

# Create custom hover text for airports
hover_text_airports = [f"<b>{airport}</b><br>" +
                       f"Delay Rate: {rate:.1%}<br>" +
                       f"<i>({rate*100:.1f} out of 100 flights delayed)</i>"
                       for airport, rate in zip(top_airports.index[:10], 
                                               top_airports.values[:10])]

fig_airports.add_trace(go.Bar(
    x=top_airports.values[:10],
    y=top_airports.index[:10],
    orientation='h',
    marker=dict(color=top_airports.values[:10], colorscale='Blues'),
    name='Airports',
    hovertemplate='%{hovertext}<extra></extra>',  # Custom hover template
    hovertext=hover_text_airports
))

# ----- Dropdown buttons for top N airports -----
buttons_airports = []
max_airports = min(20, len(top_airports))

for n in range(1, max_airports + 1):
    # Create hover text for each dropdown option
    hover_text_n = [f"<b>{airport}</b><br>" +
                   f"Delay Rate: {rate:.1%}<br>" +
                   f"<i>({rate*100:.1f} out of 100 flights delayed)</i>"
                   for airport, rate in zip(top_airports.index[:n], 
                                           top_airports.values[:n])]
    
    buttons_airports.append(dict(
        label=f"Top {n}",
        method="update",
        args=[{"x": [top_airports.values[:n]],
               "y": [top_airports.index[:n]],
               "marker.color": [top_airports.values[:n]],
               "hovertext": [hover_text_n]},
              {"title": f"Top {n} Airports by Delay Rate",
               "xaxis": {"title": "Delay Rate (%)", "tickformat": ".1%"},
               "yaxis": {"title": "Airport", "autorange": "reversed"}}]
    ))

fig_airports.update_layout(
    title="Top 10 Airports by Delay Rate",
    xaxis=dict(title="Delay Rate (%)", tickformat=".1%"),  # Format x-axis as percentage
    yaxis=dict(title="Airport", autorange="reversed"),
    plot_bgcolor='lightgray',
    paper_bgcolor='lightgray',
    width=1400,        
    height=400,
    hoverlabel=dict(
        bgcolor="white",
        font_size=14,
        font_family="Arial"
    ),
    updatemenus=[dict(
        active=9,  # default top 10
        buttons=buttons_airports,
        direction="down",
        showactive=True,
        x=0,
        y=1.15
    )]
)

# ----- Display the figures -----
fig_airlines.show()
fig_airports.show()

# ----- SAVE BOTH PLOTS AS HTML FILES -----

# Save airlines plot as HTML
fig_airlines.write_html("visualizations/airlines_delays.html")
print("Saved interactive HTML: visualizations/airlines_delays.html")

# Save airports plot as HTML
fig_airports.write_html("visualizations/airports_delays.html")
print("Saved interactive HTML: visualizations/airports_delays.html")



Saved interactive HTML: visualizations/airlines_delays.html
Saved interactive HTML: visualizations/airports_delays.html


In [100]:
# ---------------------------------
# Pie chart visualization
# ---------------------------------


# Create directory for visualizations if it doesn't exist
os.makedirs("visualizations", exist_ok=True)

# ----- Define delay count columns based on your data -----
# Update these column names to match your dataset
delay_count_columns = ['carrier_ct', 'weather_ct', 'nas_ct', 'security_ct', 'late_aircraft_ct']

# ----- Sum total delays per type -----
delay_type_totals = df_clean[delay_count_columns].sum().reset_index()
delay_type_totals.columns = ['Delay Type', 'Total']

# ----- Calculate percentages for hover text -----
total_all_delays = delay_type_totals['Total'].sum()
delay_type_totals['Percentage'] = (delay_type_totals['Total'] / total_all_delays * 100)

# ----- Create more descriptive labels -----
# Map technical column names to user-friendly descriptions
delay_descriptions = {
    'carrier_ct': 'Carrier/Airline Issues',
    'weather_ct': 'Weather Conditions',
    'nas_ct': 'National Airspace System',
    'security_ct': 'Security Issues',
    'late_aircraft_ct': 'Late Aircraft Arrival',
}

# Apply friendly names if available, otherwise clean up the original names
delay_type_totals['Display Name'] = delay_type_totals['Delay Type'].apply(
    lambda x: delay_descriptions.get(x, x.replace('_', ' ').title())
)

# ----- Create custom hover template -----
delay_type_totals['hover_text'] = delay_type_totals.apply(
    lambda row: (
        f"<b>{row['Display Name']}</b><br>"
        f"<br>"
        f"Total Delays: <b>{row['Total']:,.0f}</b><br>"
        f"Percentage: <b>{row['Percentage']:.1f}%</b><br>"
        f"<br>"
        f"<i>This represents {row['Percentage']:.1f}% of all recorded delays</i><br>"
        f"<i>({row['Total']:,.0f} out of {total_all_delays:,.0f} total delays)</i>"
    ), axis=1
)

# ----- Create interactive pie chart -----
fig_pie = px.pie(
    delay_type_totals,
    names='Display Name',  # Use friendly names for display
    values='Total',
    hole=0.3,  # donut chart
    custom_data=['hover_text']  # Pass custom hover data
)

# ----- Update traces for better visualization -----
fig_pie.update_traces(
    textinfo='percent+label',  # Show label and percentage
    textposition='inside',
    textfont_size=12,
    pull=[0.05]*len(delay_type_totals),  # Slightly separate all slices
    hovertemplate='%{customdata[0]}<extra></extra>',  # Use custom hover template
    marker=dict(
        line=dict(color='white', width=2)  # Add white borders between slices
    )
)

# ----- Update layout for better visibility -----
fig_pie.update_layout(
    title=dict(
        text='Proportion of Delay Types<br><sub>Distribution of flight delays by cause</sub>',
        font=dict(color='black', size=20),
        x=0.5,  # Center title
        y=0.95  # Moved down from 0.98 for more space from top
    ),
    plot_bgcolor='lightgray',
    paper_bgcolor='lightgray',
    width=1400,  # Adjusted for better proportions
    height=400,  # Increased height for better visibility
    hoverlabel=dict(
        bgcolor="white",
        font_size=13,
        font_family="Arial",
        bordercolor="gray"
    ),
    showlegend=True,
    legend=dict(
        orientation="v",  # Vertical legend
        yanchor="middle",
        y=0.5,
        xanchor="left",
        x=1.02,
        bgcolor="rgba(255,255,255,0.8)",
        bordercolor="gray",
        borderwidth=1
    ),
    margin=dict(t=100, b=60, l=40, r=180)
)

# ----- Add annotation with total count -----
fig_pie.add_annotation(
    text=f"Total Delays Analyzed: {total_all_delays:,.0f}",
    xref="paper", yref="paper",
    x=0.5, y=-0.1,
    showarrow=False,
    font=dict(size=11, color="gray"),
    xanchor='center'
)

# ----- Display the figure -----
fig_pie.show()


# ----- SAVE PIE CHART AS HTML FILE -----

fig_pie.write_html("visualizations/delay_types_pie.html")
print(" Saved interactive HTML: visualizations/delay_types_pie.html")


# ----- Print summary statistics -----
print("\n Delay Type Summary:")
print("-" * 60)
for _, row in delay_type_totals.iterrows():  # Fixed the syntax error here
    print(f"{row['Display Name']:30s}: {row['Total']:>10,.0f} ({row['Percentage']:>5.1f}%)")
print("-" * 60)
print(f"{'TOTAL':30s}: {total_all_delays:>10,.0f} (100.0%)")
print("\n Hover over pie slices for detailed information!")




 Saved interactive HTML: visualizations/delay_types_pie.html

 Delay Type Summary:
------------------------------------------------------------
Carrier/Airline Issues        :     53,706 ( 31.5%)
Weather Conditions            :      4,824 (  2.8%)
National Airspace System      :     54,101 ( 31.7%)
Security Issues               :        459 (  0.3%)
Late Aircraft Arrival         :     57,386 ( 33.7%)
------------------------------------------------------------
TOTAL                         :    170,477 (100.0%)

 Hover over pie slices for detailed information!


# Conclusion and Insights

## Key Findings from Airline Delay Analysis

### 1. Overall Performance
- **Total Flights:** 997,120  
- **Delayed Flights:** 170,477  
- **Overall Delay Rate:** 17.10%  
- **Interpretation:** Roughly 1 in 6 flights experienced a delay  

### 2.  Carrier Performance Ranking
- **Highest Delay Rate:** JetBlue Airways (27.35%)  
- **Other high-delay carriers:** Allegiant Air (26.43%), ExpressJet (24.84%)  
- **Lowest among top 10:** United Airlines (16.53%)  
- **Major national carriers** (United, Southwest, Alaska): 16–18% delay rate  

### 3.  Airport Performance Ranking
- **Worst Performer:** Concord, NC (Concord Padgett Regional) – 41.86%  
- **Other high-delay airports:** Redding, CA (32.26%), Bakersfield, CA (31.46%)  
- Most high-delay airports are small/regional with limited capacity  
- Larger hubs maintain lower, more stable delay rates  

### 4. Delay Distribution Patterns
- Delays are heavily concentrated in smaller carriers and airports  
- Regional airports with limited infrastructure/weather exposure face extreme rates  
- JetBlue and Allegiant alone contribute disproportionately to total delays  

### 5.  Operational Insights
- Smaller/low-cost carriers face higher operational challenges vs. major airlines  
- Hub airports manage large volumes better than small airports with fewer flights  
- Delay rates vary widely: worst airports/airlines are nearly double the system average  

### 6.  Recommendations for Improvement
- Focus on operational reliability for regional/low-cost carriers (JetBlue, Allegiant)  
- Infrastructure/ATC support for regional airports with 30%+ delay rates  
- Prioritize predictive scheduling and turnaround efficiency  
- Investigate geographic/weather-specific mitigation strategies (e.g., Aspen)  

### 7. Areas for Further Analysis
- Seasonal delay patterns (summer vs. winter)  
- Time-of-day effects (morning vs. evening flights)  
- Route-specific performance (origin–destination pairs)  
- Delay cause breakdown (carrier vs. weather vs. NAS vs. late aircraft)  
- Cost and passenger impact of delays at high-delay airports  

---

## Summary
The overall U.S. flight delay rate is **17.1%**, but certain airlines and airports significantly exceed this benchmark.  
JetBlue, Allegiant, and regional airports like Concord and Redding drive a disproportionate share of delays.  
Improvements in carrier operations and regional airport infrastructure could substantially reduce system-wide disruptions.
